Completion fixes

lorcanmooney opened the pull request:
- Fixes #4585
- ~Fixes #13161~ Cherry-picked and merged by @rchande
- Fixes a regression in VB where commiting a generic symbol with `(` wouldn't truncate the insertion text
- Fixes a handful of issues around escaping identifiers
  - In C#, `await` identifiers weren't being escaped in async methods
  - In VB, `Await` identifiers were always escaped, even when preceded by `.`
  - ~~Contextual keyword identifiers in symbol display strings weren't being escaped~~ Opened #13610 instead
- Fixes #17269 - use `ImmutableArray.Empty` instead of `Nothing`.

lorcanmooney commented on src/Features/Core/Portable/Completion/Providers/AbstractSymbolCompletionFormat.cs:
```diff
@@ -0,0 +1,153 @@
+            return insertionText;
+        }
+
+        private ImmutableArray<SymbolDisplayPart> Escape(SyntaxContext context, ImmutableArray<SymbolDisplayPart> parts)
```
This is a second pass to pick up contextual keyword identifiers (which don't seem to be handled by the compiler layer). This isn't ideal, and should probably be handled in a more central location. The symbol display service seems like an option, but that would still be act as a second pass. I'm open to suggestions...

lorcanmooney replied:
After further thought, I think it might be best to open a separate issue for this, so it can be considered independently. I'll back our the product changes, but leave the tests and mark them as skipped.

rchande commented:
@lorcanmooney Looks like you have some conflicts--mind fixing them up?

lorcanmooney commented:
@rchande Done. I've rebased the changes off master.

lorcanmooney commented:
Something seems to be wrong with the tests - they've been stuck at "Waiting for status to be reported" for almost a day now. Closing and reopening to see if it resets them.

lorcanmooney commented:
@dotnet-bot test perf_correctness_prtest please

Previous run failed because of #14090.

CyrusNajmabadi commented on src/EditorFeatures/CSharpTest/Completion/CompletionProviders/ObjectCreationCompletionProviderTests.cs:
```diff
@@ -437,5 +438,147 @@ void M1()
+    private NS.A _a = new NS.A<
+}
+";
+            await VerifyProviderCommitAsync(before, "NS.A", after, '<', string.Empty);
```
I would not do any commit tests in the Provider test framework.  Commit tests here are very busted as they don't actually invoke our real commit infrastructure.  Commit tests shuold be in the CompletionCommandHandlerTests.  @rchande can you help port these?

CyrusNajmabadi replied:
Still needs to be in CompletionCommandHandlerTests

lorcanmooney replied:
Right. I haven't made any of those changes yet. I was just looking at the options today. I'll try to move the tests over next week.

CyrusNajmabadi commented on src/EditorFeatures/Test/Completion/AbstractCompletionProviderTests.cs:
```diff
@@ -62,10 +59,39 @@ internal CompletionServiceWithProviders GetCompletionService(Workspace workspace
+            // If there are options to be set, the fixture cannot be used,
+            // since the options may impact the behaviour of other tests
+            dispose = true;
+            return CreateWorkspaceAsync(string.Empty)
```
this shoudl just be awaited.  no need for continuewith.

CyrusNajmabadi commented on src/EditorFeatures/Test/Completion/AbstractCompletionProviderTests.cs:
```diff
@@ -260,21 +290,34 @@ protected virtual bool CompareItems(string actualItem, string expectedItem)
-            var document1 = await WorkspaceFixture.UpdateDocumentAsync(code, sourceCodeKind);
-            await CheckResultsAsync(document1, position, expectedItemOrNull, expectedDescriptionOrNull, usePreviousCharAsTrigger, checkForAbsence, expectedGlyph, matchPriority);
+            bool dispose;
+            var workspace = await GetWorkspaceAsync(out dispose, options: options);
```
Is this just so we restore teh workspace back to its original state?  if so, we should capture the options before, then set the options, then restore them.

CyrusNajmabadi commented on src/EditorFeatures/VisualBasicTest/Completion/CompletionProviders/ObjectCreationCompletionProviderTests.vb:
```diff
@@ -99,5 +100,132 @@ End Namespace
+End Class
+]]></code>.Value
+
+            Await VerifyProviderCommitAsync(text, "Await", expected, Nothing, "")
```
same comment about commit tests.  These should never be done inside provider tests.  Always in the CommandHandler tests.

CyrusNajmabadi commented:
Do we have tests that verify that completing with tab/space/enter never put in the "(Of...)" part?

rchande commented:
@lorcanmooney We're trying to shut down for a release and are only taking selective changes to reduce risk. However I'd love to get in your fix for https://github.com/dotnet/roslyn/issues/13161. Can you extract out just the changes that pertain to that bug and rebase the "dev15-rc" branch? If you don't have time to do it in the next day, I'll go ahead and extract just those changes myself 😄 .

lorcanmooney commented:
I'm afraid I won't have time to work on this until the weekend. Feel free to take out whatever changes you need.

rchande commented:
Will do, thanks.

lorcanmooney commented:
@CyrusNajmabadi can you take a look at my changes to `AbstractCompletionProviderTests`, specifically [this commit](https://github.com/dotnet/roslyn/pull/13430/commits/45364f8e8186858d24446ef221e60ccb765c94ad). This seems like an unreasonable amount of work just to save a snapshot, but I simply couldn't find another way to make it work.

There are a couple of surprising things I found in the options API:
- The implication of the [doc comments for `WorkspaceOptionSet`](https://github.com/dotnet/roslyn/blob/c723ac13c874b7dd867cca79324535e42fcd5a45/src/Workspaces/Core/Portable/Options/WorkspaceOptionSet.cs#L10). This implies you can never reliably use any `OptionSet` as a snapshot of the current state, since it would be impossible to guarantee all options have been "observed".
- Reading the `Options` property [always returns a new `OptionSet` instance](https://github.com/dotnet/roslyn/blob/c723ac13c874b7dd867cca79324535e42fcd5a45/src/Workspaces/Core/Portable/Options/OptionServiceFactory.cs#L118), losing track of which options have been "observed".

Maybe I'm missing something?

CyrusNajmabadi commented on src/EditorFeatures/Test/Completion/AbstractCompletionProviderTests.cs:
```diff
@@ -188,60 +185,64 @@ protected virtual bool CompareItems(string actualItem, string expectedItem)
             string markup, string expectedItem, string expectedDescriptionOrNull = null,
-            SourceCodeKind? sourceCodeKind = null, bool usePreviousCharAsTrigger = false)
+            SourceCodeKind? sourceCodeKind = null, bool usePreviousCharAsTrigger = false,
+            IEnumerable<KeyValuePair<OptionKey, object>> options = null)
```
Is there a reason these aren't more simply just ```IDictionary<OptionKey, object>```s?

lorcanmooney replied:
Not that I remember. I'll change them.

CyrusNajmabadi commented on src/EditorFeatures/Test/Completion/AbstractCompletionProviderTests.cs:
```diff
@@ -260,21 +261,57 @@ protected virtual bool CompareItems(string actualItem, string expectedItem)
-            if (await CanUseSpeculativeSemanticModelAsync(document1, position))
+            OptionSet defaultOptions = null;
+
+            if (options != null && options.Any())
```
no need for the .Any check.

CyrusNajmabadi commented:
@jasonmalinowski  To look at this.  This appears to be an 'interesting' quirk of options, especially workspace options.  Not sure if there's something better.

lorcanmooney commented:
@CyrusNajmabadi I've made the requested changes.

lorcanmooney commented:
https://ci.dot.net/job/dotnet_roslyn/job/master/job/windows_build_correctness_prtest/56/

```
Running RepoUtil
Verifying project.json contents
Verifying RepoData.json
Error: Fixed package Microsoft.Build - 14.3.0 is not used anywhere
Error: Fixed package Microsoft.Build.Tasks.Core - 14.3.0 is not used anywhere

Error! RepoUtil verification failed.
```

Ok... I'm fairly sure that wasn't me.

lorcanmooney commented:
@dotnet-bot retest windows_build_correctness_prtest please

Pilchie commented:
Tagging @jaredpar for the error above.

jaredpar commented:
That error in RepoUtil was introduced on Friday and fixed on Sunday by a different change.  If you merge / rebase against master it should go away.

lorcanmooney commented:
Looks like an xunit failure:

```
System.NullReferenceException: Object reference not set to an instance of an object.
   at Xunit.Sdk.TestRunner`1.<RunAsync>d__43.MoveNext() in C:\BuildAgent\work\cb37e9acf085d108\src\xunit.execution\Sdk\Frameworks\Runners\TestRunner.cs:line 150
```

@dotnet-bot retest windows_release_unit32_prtest please

lorcanmooney commented:
@dotnet-bot retest windows_release_unit32_prtest please

lorcanmooney commented:
So, now that VS2017 is out there, what are odds of making some progress on this? I'm can bring it up-to-date but, frankly, I don't want to waste my time resolving the _next_ three months of merge conflicts. :-/

Pilchie commented:
Sorry about that @lorcanmooney  - if you get it up to date, we'll happily merge it for 15.3 ASAP.

CyrusNajmabadi commented:
I'm looking at resolving the merge conflicts for this now.

CyrusNajmabadi commented:
@lorcanmooney I've provided a PR that merges master into your branch for you.  I've resolved all the completion conflicts between your changes and my changes.

lorcanmooney commented:
The windows_release_vs-integration_prtest failure is #17816.

I'm get `Access Denied` when I try to look at the windows_eta_open_prtest results. Can someone let me know if it's something I need to investigate?

@jasonmalinowski There's still an open question in the comments above about how the `WorkspaceOptionSet` works. Maybe you could take a quick look at the code?

jasonmalinowski commented:
@lorcanmooney Sorry I missed the original question. I must admit I don't understand what the code is trying to achieve or work around. What were you running into?

lorcanmooney commented:
@jasonmalinowski These tests share a workspace via a class fixture. Some of the tests validate behaviour with non-default options. The goal of my changes was to apply these non-default values, then restore the options to the original state once the test is complete.

The `OptionSet` is presented as what looks like a standard immutable API, so my plan was to read `Workspace.Options` and store it for later. In the meantime I would use that snapshot to create a new instance with the non-default values needed for the test, and apply them to the workspace (this part worked fine). Once the test was done, my attempts to apply the original `OptionSet` to the workspace never seemed to have any effect - the non-default options would stay in effect and cause subsequent tests to fail. The reason I found was that `WorkspaceOptionSet` is not actually a snapshot, but a cache only of "observed" values. Before the test, no options had been observed, so restoring the original instance just restored an empty cache - reads still went through to the underlying `IOptionService` which still contained the non-default values.

My workaround for this was to "observe" all the options _before_ I changed them. That way, I could get an `OptionSet` primed with the default values, so applying it afterwards would actually change something. This was frustrated by always getting an empty `WorkspaceOptionSet` from `Workspace.Options`, which was effectively throwing away any of the default values I'd previously observed.

tl;dr: Options didn't work at all like I expected. Were my expectations unreasonable, and are my workarounds the correct approach?

jasonmalinowski commented:
@lorcanmooney I would say the answer is "that's not a scenario we really intended to support, but yeah, that's baffling behavior indeed". Your workaround is acceptable. Feel free to file a bug and mention the bug URL in the workaround so we can think of something to do here. I think the fixes might be simple to do.

lorcanmooney commented:
Thanks @jasonmalinowski. I've opened #13430 and updated the work-around to reference it.

Pilchie commented:
@CyrusNajmabadi Can you bring this PR current again, and let's get it merged.

@lorcanmooney sorry for forgetting this for so long 😦

Neme12 commented on src/EditorFeatures/CSharpTest/Completion/CompletionProviders/ObjectCreationCompletionProviderTests.cs:
```diff
@@ -438,6 +439,58 @@ void M1()
+}";
+
+            await VerifyItemExistsAsync(markup, "A<Int32>",
+                options: Option(CodeStyleOptions.PreferIntrinsicPredefinedTypeKeywordInDeclaration, LanguageNames.CSharp, false));
```
I don't quite understand this. Why do we need to have options involved in completion at all? This doesn't seem like a good behavior to me. I would expect this to respect what you actually typed in the declaration as opposed to checking your style preferecnes:
```c#
A<int> a = new $$ // I would expect A<int> to show up here
A<Int32>a = new && // I would expect A<Int32> to show up here
```

lorcanmooney replied:
Today it doesn't respect either what I type _or_ the preferences I've set. In most cases I would expect the desire is to have both the variable and creation styles match the current preferences.

Taking it further would probably require a good chunk of extra work for the minimizing process to produce the correct result, and dealing with cases where the target of the assignment doesn't have any available source code, or where that source code style preferences don't match the current document. Not something I'd consider worth the effort.

Neme12 replied:
My problem with this is that completion is forcing you to a style change even though you already explicitly typed what you typed in a different style for whatever reason. Completion should respect that first.

>  I would expect the desire is to have both the variable

Then write it that way in the first place :smiley: If you don't, completion shouldn't be forcing you to.

Neme12 replied:
> where the target of the assignment doesn't have any available source code

could you clarify what you mean by that?

Neme12 replied:
@CyrusNajmabadi what do you think about this?

CyrusNajmabadi replied:
THere are two situations to care about:

1. ```some_expr = new$$```  and
2. ```ExplicitType t = new $$```

Yes.  In the second case, it seems reasonable to probably just suggest whatever type was explicit provided on the left.  This code, however, is clearly trying to make sure that case 1 works properly.  i.e. if you had an expression that imported a type (say ```List<int>```) that "new" will respect your wishes.

CyrusNajmabadi replied:
> I don't quite understand this. Why do we need to have options involved in completion at all? 

Because sometimes you have types that you don't know how they were originally declared (this is common) and you want to ensure that any generation you do follows the style that hte user has explicitly stated they want.

Neme12 commented on src/EditorFeatures/CSharpTest/Completion/CompletionProviders/ObjectCreationCompletionProviderTests.cs:
```diff
@@ -438,6 +439,58 @@ void M1()
+        }
+
+        [Fact(Skip = "https://github.com/dotnet/roslyn/issues/13610"), Trait(Traits.Feature, Traits.Features.Completion)]
+        public async Task EscapeAwaitIdentifierInAsyncMethod()
```
Skip?
In the PR you said this was fixed:
> Fixes a handful of issues around escaping identifiers
In C#, await identifiers weren't being escaped in async methods

lorcanmooney replied:
Some cases were fixed, some remain broken. Those that insert "minimal" text are still broken because of #13610, while those that insert non-minimal strings are fixed by these changes.

Neme12 commented on src/EditorFeatures/CSharpTest/Completion/CompletionProviders/ReferenceDirectiveCompletionProviderTests.cs:
```diff
@@ -33,12 +35,12 @@ protected override bool CompareItems(string actualItem, string expectedItem)
             string code, int position, string expectedItemOrNull, string expectedDescriptionOrNull,
             SourceCodeKind sourceCodeKind, bool usePreviousCharAsTrigger, bool checkForAbsence,
-            int? glyph, int? matchPriority, bool? hasSuggestionItem)
+            int? glyph, int? matchPriority, bool? hasSuggestionItem, IDictionary<OptionKey, object> options)
```
As I said, I'm not too excited about options being involved in completion. What is the reason for that?

lorcanmooney replied:
> Could you explain why you think it's necessary?

I told VS to generate code according to a particular style. I expect my preferences to be honoured, and they are, with the exception of code generated by the completion providers. To fix that, the providers need to know what the current set of preferences are.

Neme12 replied:
I don't see a reason to do any complciated code generation here based on your preferred style.
We offer the name even if it doesn't exist:
![image](https://user-images.githubusercontent.com/11444821/37253756-22bd9446-2536-11e8-8f87-81df33d25cea.png)

Neme12 replied:
Is there a reason why this couldn't be entirely syntax based?

lorcanmooney replied:
You're focusing on one very specific, simple case. What would you expect to see for an argument to a method call?

```csharp
Foo(new $$)
```

Or an assignment to a property?

```csharp
x.Foo = new $$
```

How would you expect this to behave when the methods, properties and types involved are defined in a referenced assembly (so you don't know what the original code style is), or what happens if they're defined in a project, but that project has an editorconfig style that conflicts with the editorconfig used for the document you're typing in. Also consider that the type your instantiating may not be in any of your imported namespaces, so an extra resolution and minimizing process is applied.

I'm not saying it's impossible to do what you want, but there are a lot of cases where an obvious raw textual syntax is not readily available. And that's just considering the object-creation provider. In those cases you have to pick a code style, and that style should match what the user has set in their preferences.

Now, if you want to build on that for the special cases where a specific syntax _is_ available, then go ahead, I certainly wouldn't object.

Neme12 replied:
@lorcanmooney Alright thanks for the explanation, it makes more sense now. I thought this completion was specific to declaring a variable. I agree it's a lot more complicated if we're offering this everywhere. Sorry for bothering you

Neme12 commented on src/EditorFeatures/CSharpTest/Completion/CompletionProviders/SymbolCompletionProviderTests.cs:
```diff
@@ -9129,9 +9129,40 @@ void M()
+        a$$
+    }
+}";
+            await VerifyItemExistsAsync(markup, "@await");
```
this test should probably also have `await VerifyItemIsAbsentAsync(markup, "await");`

lorcanmooney commented:
@Pilchie @CyrusNajmabadi This PR has been open for over a year and a half now, and despite addressing al the feedback, still hasn't been merged. So, I've brought it up to date ... again. If there's a problem with something I'm doing here please let me know, otherwise I'd like to make some progress.

Neme12 commented on src/EditorFeatures/CSharpTest/Completion/CompletionProviders/SymbolCompletionProviderTests.cs:
```diff
@@ -9129,9 +9129,40 @@ void M()
+        a$$
+    }
+}";
+            await VerifyItemIsAbsentAsync(markup, "@await");
```
same here, you could verify that `@await` doesn't exist but `await` does

Neme12 commented on src/Workspaces/CSharp/Portable/Extensions/ContextQuery/CSharpSyntaxContext.cs:
```diff
@@ -255,11 +255,10 @@ public static CSharpSyntaxContext CreateContext_Test(SemanticModel semanticModel
-            // method or not.
-            return false;
+            var methodDeclaration = targetToken.GetAncestor<MethodDeclarationSyntax>();
+            return methodDeclaration != null && methodDeclaration.Modifiers.Any(SyntaxKind.AsyncKeyword);
```
what about async local functions? lambdas? anonymous methods?

lorcanmooney replied:
Hmm, good point...

Neme12 replied:
please add tests for that too

lorcanmooney replied:
So, this is already broken for `Async Function` in VB, and since I based the C# version on that, it inherits the problem. Async local functions and anonymous delegates<sup>*</sup> seem to reset the `async`-ness (unless of course they too are `async`), whereas lambdas inherit the `async`-ness of the nearest Method, local function, anonymous delegate or async lambda. Or something like that.

It really feels like this info should come from the compiler, instead of trying replicate the calculation. In any case, this is enough work for a separate issue, and I'm not making it any worse, so #wontfix :-)

<sub><sup>*</sup> I didn't even know `async delegate` was a thing.</sub>

Neme12 replied:
@lorcanmooney 
> whereas lambdas inherit the async-ness of the nearest Method

Yes, I checked that too. It seems really weird that the behavior is different between lambdas / anonymous methods. Might be a bug?

With lambdas, either making the lambda itself or the containing method does the trick. At least that's what syntax highlgihting is telling me.

> In any case, this is enough work for a separate issue, and I'm not making it any worse, so #wontfix :-)

I agree this is a very minor bug, so you don't have to fix it. But I think the case inside a method is pretty minor as well. 😃

Neme12 replied:
@lorcanmooney just so you know, I filed an issue for the behavior with lambdas: #25410 because it is most likely a bug. That would make the fix easier than you thought, but it's ok if you still don't want to do it. I might do it later myself.

CyrusNajmabadi commented:
> This PR has been open for over a year and a half now, and despite addressing all the feedback, still hasn't been merged. 

Sorry, i have no control over that :-/ .  I'll take a look at the feedback and see if i'm still ok.  But, unfortunately, some things do slip through teh cracks.  I recommend when that happens pinging a little sooner.  Also try @sharwell as he really seems to work hard on community PRs.

CyrusNajmabadi commented on src/Features/Core/Portable/Completion/Providers/AbstractPartialTypeCompletionProvider.cs:
```diff
@@ -16,8 +17,16 @@ namespace Microsoft.CodeAnalysis.Completion.Providers
+                throw new ArgumentNullException(nameof(format));
+            }
+
+            _format = format;
```
nit: consider:  ```_format = format ?? throw```

CyrusNajmabadi commented on src/Features/Core/Portable/Completion/Providers/AbstractSymbolCompletionFormat.cs:
```diff
@@ -0,0 +1,101 @@
+
+        public string GetInsertionTextAtInsertionTime(CompletionItem item, char? ch)
+        {
+            string insertionText = SymbolCompletionItem.GetInsertionText(item);
```
nit: generally we prefer var in the LS.

Pilchie commented:
Tagging @sharwell and @jinujoseph

jinujoseph commented:
@ivanbasov , Can you review this with context with all the changes that has happened in completion

lorcanmooney commented:
This PR has quite literally been around for years, and I've already spent more time resolving merge conflicts than the fix will ever save me. Frankly, I've no intention of wasting any more time here.
